src/charm.py

```diff
@@ -919,6 +880,47 @@ def _generate_prometheus_config(self) -> bool:
         logger.info("Pushed new configuration")
         return True
 
+    def _process_tls_config(self, job) -> Dict:
+        certs = {}
+        if (tls_config := job.get("tls_config", {})) or job.get("scheme") == "https":
+            # Certs are transferred over relation data and need to be written to files on disk.
+            # CA certificate to validate the server certificate with.
+
+            # If the scrape job has a TLS section but no "ca_file", then use ours, assuming
+            # prometheus and all scrape jobs are signed by the same CA.
+            if ca_file := tls_config.get("ca_file") or self.cert_handler.ca:
+                # TODO we shouldn't be passing CA certs over relation data, because that
+                #  reduces to self-signed certs. Both parties need to separately trust the CA
+                #  instead.
+                filename = f"{PROMETHEUS_DIR}/{job['job_name']}-ca.crt"
+                certs[filename] = ca_file
+                job["tls_config"] = {**tls_config, **{"ca_file": filename}}
+            else:
+                # The tls_config section is present, but we don't have any CA certs
+                logger.warning(
+                    "The scrape job '%s' has a tls_config section specified, but no CA certs "
+                    "are available.",
+                    job["job_name"],
+                )
+
+            # Certificate and key files for client cert authentication to the server.
+            if (cert_file := tls_config.get("cert_file")) and (
+                key_file := tls_config.get("key_file")
+            ):
+                filename = f"{PROMETHEUS_DIR}/{job['job_name']}-client.crt"
+                certs[filename] = cert_file
+                job["tls_config"]["cert_file"] = filename
+                filename = f"{PROMETHEUS_DIR}/{job['job_name']}-client.key"
+                certs[filename] = key_file
+                job["tls_config"]["key_file"] = filename
+            elif "cert_file" in tls_config or "key_file" in tls_config:
+                raise ConfigError(
+                    'tls_config requires both "cert_file" and "key_file" if client '
+                    "authentication is to be used"
+                )
+
+        return {"job": job, "certs": certs}
```

Comment:
Should we simply `return job, certs` here, so that we could `processed_job, extracted_certs = self._process_tls_config(job)` above?

Reply:
Yes, it's a little bit golanger, but I can live with that 😄